gtk/src/toga_gtk/widgets/progressbar.py

```diff
@@ -1,45 +1,86 @@
-from ..libs import GLib, Gtk
+import asyncio
+
+from travertino.size import at_least
+
+from ..libs import Gtk
 from .base import Widget
 
-PROGRESSBAR_TICK_INTERVAL = 100  # ms per tick
+# Implementation notes
+# ====================
+#
+# * GTK ProgressBar doesn't have any concept of running; we track the running
+#   status for API compliance.
+#
+# * Indeterminate GTK ProgressBars need to be manually animated; when an
+#   indeterminate progress bar is started, we add a background task to do this
+#   pulse animation every 100ms.
+#
+# * GTK ProgressBar uses 0-1 floating point range. We track the Toga max value
+#   internally for scaling purposes.
+
+
+async def pulse(progressbar):
+    "A background task to animate running indeterminate progress bars"
+    while True:
+        progressbar.native.pulse()
+        await asyncio.sleep(0.1)
 
 
 class ProgressBar(Widget):
     def create(self):
         self.native = Gtk.ProgressBar()
 
-    def _render_disabled(self):
-        self.native.set_fraction(0)
+        self._max = 1.0
+        self._running = False
+        self._task = None
+
+    def is_running(self):
+        return self._running
+
+    def get_value(self):
+        if self.get_max() is None:
+            return None
+        return self.native.get_fraction() * self._max
 
     def set_value(self, value):
-        self.native.set_fraction(self.interface.value / self.interface.max)
+        self.native.set_fraction(value / self._max)
+
+    def get_max(self):
+        return self._max
+
+    def _start_indeterminate(self):
+        self._task = asyncio.create_task(pulse(self))
+
+    def _stop_indeterminate(self):
+        if self._task:
+            self._task.cancel()
+            self._task = None
+            self.native.set_fraction(0.0)
 
     def set_max(self, value):
-        if not self.interface.enabled:
-            self._render_disabled()
+        self._max = value
```

Comment:
It looks like changing the max in determinate mode will not update the progress bar's position. There should be a testbed test for this, perhaps using a `probe.position` property like the Slider.

Reply:
Good catch. This also revealed an edge case with iOS; changing the max cleared the value.